File: TestStand/Examples/TestStand API/Executing Sequences Using API/CVI/ExecutingSequencesUsingAPI.c
//==============================================================================
//
// Title:       Executing Sequences Using API -LabWindows_CVI.c 
//
// Description:	This file holds the three functions used in the Executing
//				Sequences Using API - LabWindows_CVI for TestStand.
//
//==============================================================================

#include "tsutil.h"

#define ERROR_MESSAGE_SIZE 1024

//==============================================================================
//
//	This function executes a sequence with no process model.
//
//	Inputs:
//		seqContextCVI:			The sequence context.
//		sequenceFileToRun:		The name of the sequence file to run located in the TestStand search directories.
//		sequenceNameParameter:	The name of the sequence to run within the sequence file.
//
//	Outputs:
//		errorOccured:			TRUE or FALSE to represent if an error has occured or not.
//		errorCode:				The error code if errorOccured is TRUE.
//		errorMsg:				The error message if errorOccured is TRUE.
//
void __declspec(dllexport) __stdcall ExecuteWithNoProcessModel(CAObjHandle seqContextCVI, char sequenceFileToRun[1024],
		char sequenceNameParameter[1024], short *errorOccurred, long *errorCode, char errorMsg[ERROR_MESSAGE_SIZE])
{
	CAObjHandle engineHandle = 0;
	TSObj_SeqFile CurrentSeqFileHandle = 0;
	TSObj_SeqFile sequenceFileEx = 0;
	CAObjHandle executionHandle = 0;

	char *absolutePath = 0;
	VARIANT seqFileVar = CA_VariantEmpty ();
	VARIANT seqExecutionVar = CA_VariantEmpty ();
	int error = 0;
	
	ErrMsg errMsg = "";
	ERRORINFO errorInfo;

	// Get the engine handle.
	tsErrChk (TS_SeqContextGetEngine (seqContextCVI, &errorInfo, &engineHandle));

	// Get the sequence file handle.
	tsErrChk (TS_SeqContextGetSequenceFile (seqContextCVI, &errorInfo, &CurrentSeqFileHandle));
	
	// Copy the sequence file handle to a variant variable to be used in the next step. 
	CA_VariantSetObjHandle (&seqFileVar, CurrentSeqFileHandle, CAVT_DISPATCH);
	
	// Find the sequence file's absolute path based on the calling sequence file's directory.
	tsErrChk (TS_EngineFindFile (engineHandle, &errorInfo, sequenceFileToRun,
								 &absolutePath, NULL,
								 TS_FindFile_PromptHonorUserPreference,
								 TS_FindFile_AddDirToSrchList_Ask, 
								 VFALSE, seqFileVar, NULL));

	// Locate and open the sequence file containing the sequence to be executed.
	tsErrChk (TS_EngineGetSeqFileEx (engineHandle, &errorInfo, absolutePath, TS_GetSeqFile_FindFile, TS_ConflictHandler_Error, &sequenceFileEx));

	// Launch a new execution of the specified sequence.
	tsErrChk (TS_EngineNewExecution (engineHandle, &errorInfo, sequenceFileEx, sequenceNameParameter, 0, VFALSE, TS_ExecTypeMask_Normal,
									 CA_DEFAULT_VAL, CA_DEFAULT_VAL, CA_DEFAULT_VAL,&executionHandle));
	
	// Wait for the execution to end.
	CA_VariantSetObjHandle (&seqExecutionVar, seqContextCVI, CAVT_DISPATCH);
	tsErrChk (TS_ExecutionWaitForEndEx (executionHandle, &errorInfo, INFINITE, VTRUE, CA_DEFAULT_VAL, seqExecutionVar, NULL));
		
	// Release the sequence file opened previously.
	tsErrChk (TS_EngineReleaseSeqFileEx (engineHandle, &errorInfo, sequenceFileEx, 0, NULL));


Error:
	// FREE RESOURCES
	CA_FreeMemory(absolutePath);
	CA_VariantClear (&seqExecutionVar);
	CA_DiscardObjHandle (executionHandle);
	CA_DiscardObjHandle (sequenceFileEx);
	CA_VariantClear (&seqFileVar);
	CA_DiscardObjHandle (CurrentSeqFileHandle);
	CA_DiscardObjHandle (engineHandle);
	
	// If an error occurred, set the error flag to cause a run-time error in TestStand.
	if (error < 0)
	{
		*errorOccurred = TRUE;

		// SET THE ERROR CODE AND STRING
		*errorCode = error;
		strncpy(errorMsg, errMsg, ERROR_MESSAGE_SIZE);
		errorMsg[ERROR_MESSAGE_SIZE-1] = 0;
	}
}


//==============================================================================
//
//	This function executes a sequence with a process model.
//
//	Inputs:
//		seqContextCVI:			The sequence context.
//		sequenceFileToRun:		The name of the sequence file to run located in the TestStand search directories.
//		processModelName:		The name of the process model to run located in the TestStand search directoreis.
//		sequenceNameParameter:	The name of the sequence to run within the sequence file.
//
//	Outputs:
//		errorOccured:			TRUE or FALSE to represent if an error has occured or not.
//		errorCode:				The error code if errorOccured is TRUE.
//		errorMsg:				The error message if errorOccured is TRUE.
//
void __declspec(dllexport) __stdcall ExecuteWithProcessModel(CAObjHandle seqContextCVI, char sequenceFileToRun[1024], char processModelName[1024],
		char sequenceNameParameter[1024], short *errorOccurred, long *errorCode, char errorMsg[ERROR_MESSAGE_SIZE])
{
	CAObjHandle engineHandle = 0;
	TSObj_SeqFile CurrentSeqFileHandle = 0;
	TSObj_SeqFile sequenceFileEx = 0;
	TSObj_SeqFile processModelEx = 0;
	CAObjHandle executionHandle = 0;

	char *absolutePath = 0;
	VARIANT seqFileVar = CA_VariantEmpty ();
	VARIANT seqExecutionVar = CA_VariantEmpty ();
	int error = 0;
	
	ErrMsg errMsg = {'\0'};
	ERRORINFO errorInfo;

	// Get the engine handle.
	tsErrChk (TS_SeqContextGetEngine (seqContextCVI, &errorInfo, &engineHandle));
	
	// Get the sequence file handle.
	tsErrChk (TS_SeqContextGetSequenceFile (seqContextCVI, &errorInfo, &CurrentSeqFileHandle));
	
	// Copy the sequence file handle to a variant variable to be used in the next step. 
	CA_VariantSetObjHandle (&seqFileVar, CurrentSeqFileHandle, CAVT_DISPATCH);
	
	// Find the sequence file's absolute path based on the calling sequence file's directory.
	tsErrChk (TS_EngineFindFile (engineHandle, &errorInfo, sequenceFileToRun, &absolutePath, NULL, TS_FindFile_PromptEnable,
								 TS_FindFile_AddDirToSrchList_Yes, VFALSE, seqFileVar, NULL));
	
	// Locate and opens the sequence file containing the sequence to be executed.
	tsErrChk (TS_EngineGetSeqFileEx (engineHandle, &errorInfo, absolutePath, TS_GetSeqFile_FindFile, TS_ConflictHandler_Error, &sequenceFileEx));
	
	// Locate and opens the process model to be used for this execution.
	tsErrChk (TS_EngineGetSeqFileEx (engineHandle, &errorInfo, processModelName, TS_GetSeqFile_FindFile, TS_ConflictHandler_Error, &processModelEx));
	
	
	// Launch a new execution of the sequence using the process model entry point specified.
	tsErrChk (TS_EngineNewExecution (engineHandle, &errorInfo, sequenceFileEx, sequenceNameParameter, processModelEx, VFALSE, TS_ExecTypeMask_Normal,
									 CA_DEFAULT_VAL, CA_DEFAULT_VAL, CA_DEFAULT_VAL,&executionHandle));

   	// Wait for the execution to end.  
	CA_VariantSetObjHandle (&seqExecutionVar, seqContextCVI, CAVT_DISPATCH);
	tsErrChk (TS_ExecutionWaitForEndEx (executionHandle, &errorInfo, INFINITE, VTRUE, CA_DEFAULT_VAL, seqExecutionVar, NULL));
	
	// Release the process model and client sequence file opened previously.
	tsErrChk (TS_EngineReleaseSeqFileEx (engineHandle, &errorInfo, processModelEx, 0, NULL));
	tsErrChk (TS_EngineReleaseSeqFileEx (engineHandle, &errorInfo, sequenceFileEx, 0, NULL));

Error:
	// FREE RESOURCES
	CA_FreeMemory(absolutePath);
	CA_VariantClear (&seqExecutionVar);
	CA_DiscardObjHandle (executionHandle);
	CA_DiscardObjHandle (sequenceFileEx);
	CA_DiscardObjHandle (processModelEx);
	CA_VariantClear (&seqFileVar);
	CA_DiscardObjHandle (CurrentSeqFileHandle);
	CA_DiscardObjHandle (engineHandle);

	// If an error occurred, set the error flag to cause a run-time error in TestStand.
	if (error < 0)
	{
		*errorOccurred = TRUE;

		// SET THE ERROR CODE AND STRING
		*errorCode = error;
		strncpy(errorMsg, errMsg, ERROR_MESSAGE_SIZE);
		errorMsg[ERROR_MESSAGE_SIZE-1] = 0;
	}
}


//==============================================================================
//
//	This function executes a sequence with sequence parameters.
//
//	Inputs:
//		seqContextCVI:			The sequence context.
//		sequenceFileToRun:		The name of the sequence file to run located in the TestStand search directories.
//		sequenceNameParameter:	The name of the sequence to run within the sequence file.
//		numberValue:			A number to be set as a parameter and displayed to the user.
//		stringValue:			A string to be set as a parameter and displayed to the user.
//
//	Outputs:
//		errorOccured:			TRUE or FALSE to represent if an error has occured or not.
//		errorCode:				The error code if errorOccured is TRUE.
//		errorMsg:				The error message if errorOccured is TRUE.
//
void __declspec(dllexport) __stdcall ExecuteWithSequenceParameters(CAObjHandle seqContextCVI, char sequenceFileToRun[1024],
		char sequenceNameParameter[1024], double numberValue, char stringValue[1024], short *errorOccurred, long *errorCode, char errorMsg[ERROR_MESSAGE_SIZE])
{
	
	CAObjHandle engineHandle = 0;
	TSObj_SeqFile CurrentSeqFileHandle = 0;  
	TSObj_Property propertyObject = 0;
	TSObj_SeqFile sequenceFileEx = 0;
	CAObjHandle executionHandle = 0;

	char *absolutePath = 0;
	VARIANT propertyObjectVar = CA_VariantEmpty ();
	VARIANT seqFileVar = CA_VariantEmpty ();
	VARIANT seqExecutionVar = CA_VariantEmpty ();
	int error = 0;
	
	ErrMsg errMsg = "";
	ERRORINFO errorInfo;

	// Get the engine handle.
	tsErrChk (TS_SeqContextGetEngine (seqContextCVI, &errorInfo, &engineHandle));

	// Get the sequence file handle.
	tsErrChk (TS_SeqContextGetSequenceFile (seqContextCVI, &errorInfo, &CurrentSeqFileHandle));
	
	// Copy the sequence handle to a variant variable to be used in the next step. 
	CA_VariantSetObjHandle (&seqFileVar, CurrentSeqFileHandle, CAVT_DISPATCH);
	
	// Find the sequence file's absolute path based on the calling sequence file's directory.
	tsErrChk (TS_EngineFindFile (engineHandle, &errorInfo, sequenceFileToRun, &absolutePath, NULL, TS_FindFile_PromptEnable,
								 TS_FindFile_AddDirToSrchList_Yes, VFALSE, seqFileVar, NULL));

	// Locate and opens the sequence file containing the sequence to be executed.
	tsErrChk (TS_EngineGetSeqFileEx (engineHandle, &errorInfo, absolutePath, TS_GetSeqFile_FindFile, TS_ConflictHandler_Error, &sequenceFileEx));

	// Create a "container" PropertyObject and add a subproperty for each parameter of the sequence to call.
	// Note: Subproperties must be in the same order as parameters appear in the sequence.
	tsErrChk (TS_EngineNewPropertyObject (engineHandle, &errorInfo, TS_PropValType_Container, VFALSE, "Parameters", 0, &propertyObject));
	tsErrChk (TS_PropertySetValNumber (propertyObject, &errorInfo, "NumberVal", 1, numberValue));
	tsErrChk (TS_PropertySetValString (propertyObject, &errorInfo, "StringVal", 1, stringValue));

	// Launch a new execution of the sequence.     
	CA_VariantSetObjHandle (&propertyObjectVar, propertyObject, CAVT_DISPATCH);
	tsErrChk (TS_EngineNewExecution (engineHandle, &errorInfo, sequenceFileEx, sequenceNameParameter, 0, VFALSE, TS_ExecTypeMask_Normal,
									 propertyObjectVar, CA_DEFAULT_VAL, CA_DEFAULT_VAL, &executionHandle));
	
	// Wait for the execution to end.      	
	CA_VariantSetObjHandle (&seqExecutionVar, seqContextCVI, CAVT_DISPATCH);
	tsErrChk (TS_ExecutionWaitForEndEx (executionHandle, &errorInfo, INFINITE, VTRUE, CA_DEFAULT_VAL, seqExecutionVar, NULL));
	
	// Release the sequence file opened previously.
	tsErrChk (TS_EngineReleaseSeqFileEx (engineHandle, &errorInfo, sequenceFileEx, 0, NULL));


Error:
	// FREE RESOURCES
	CA_FreeMemory(absolutePath);
	CA_VariantClear (&seqExecutionVar);
	CA_DiscardObjHandle (executionHandle);
	CA_DiscardObjHandle (sequenceFileEx);
	CA_VariantClear (&propertyObjectVar);
	CA_DiscardObjHandle (propertyObject);
	CA_VariantClear (&seqFileVar);
	CA_DiscardObjHandle (CurrentSeqFileHandle);
	CA_DiscardObjHandle (engineHandle);
	
	// If an error occurred, set the error flag to cause a run-time error in TestStand.
	if (error < 0)
	{
		*errorOccurred = TRUE;

		// SET THE ERROR CODE AND STRING
		*errorCode = error;
		strncpy(errorMsg, errMsg, ERROR_MESSAGE_SIZE);
		errorMsg[ERROR_MESSAGE_SIZE-1] = 0;
	}
}

File: TestStand/Examples/TestStand API/Executing Sequences Using API/DotNet/Executing Sequences Using API/ExecutingSequencesUsingAPI/ExecutingSequencesUsingAPI.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using NationalInstruments.TestStand.Interop.API;    // TestStand Core API.  Add <TestStand>\API\DotNet\Assemblies\CurrentVersion\NationalInstruments.TestStand.Interop.API.dll to the project as a reference.
using System.Runtime.InteropServices;


//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//
// Title:       Executing Sequences Using API.cs
//
// Description: This file holds the three methods used in the Executing Sequences Using API
//              for TestSTand.
//
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

namespace ExecutingSequencesUsingAPI
{
    public class TestStandSequenceCalls
    {
        
        //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        //
        //	This method executes a sequence with no process model.
        //
        //	Inputs:
        //		seqContext:	    		The sequence context.
        //		sequenceFileToRun:		The name of the sequence file to run located in the TestStand search directories.
        //		sequenceNameParameter:	The name of the sequence to run within the sequence file.
        //
        //	Outputs:
        //		errorOccured:			TRUE or FALSE to represent if an error has occured or not.
        //		errorCode:				The error code if errorOccured is TRUE.
        //		errorMsg:				The error message if errorOccured is TRUE.
        //
        public void ExecuteWithNoProcessModel(SequenceContext seqContext, string sequenceFileToRun, string sequenceNameParameter,
            out bool errorOccurred, out int errorCode, out String errorMsg)
        {
           
            string absolutePath;
            bool userCancelled;

            errorOccurred = false;
            errorCode = 0;
            errorMsg = String.Empty;

            try
            {
                // Find the sequence file absolute path based on the calling sequence file's directory.
                seqContext.Engine.FindFile(sequenceFileToRun, out absolutePath, out userCancelled, FindFilePromptOptions.FindFile_PromptHonorUserPreference,
                    FindFileSearchListOptions.FindFile_AddDirToSrchList_No, false, seqContext.SequenceFile);

                // Locate and open the sequence file contianing the sequence to be executed.
				SequenceFile targetSeqFile = seqContext.Engine.GetSequenceFileEx(absolutePath);

                // Launch a new execution of the sequence.
				Execution currentExecution = seqContext.Engine.NewExecution(targetSeqFile, sequenceNameParameter, null, false, 0);

                // Wait for the execution to end.
                currentExecution.WaitForEndEx(-1, true, Type.Missing, seqContext);

                // Release the sequence file opened previously.
                seqContext.Engine.ReleaseSequenceFileEx(targetSeqFile, 0);


            }
            catch (COMException e)
            {
                errorOccurred = true;
                errorMsg = e.Message;
                errorCode = e.ErrorCode;
            }
        }

        //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        //
        //	This method executes a sequence with a process model.
        //
        //	Inputs:
        //		seqContext: 			The sequence context.
        //		sequenceFileToRun:		The name of the sequence file to run located in the TestStand search directories.
        //		processModelName:		The name of the process model to run located in the TestStand search directoreis.
        //		sequenceNameParameter:	The name of the sequence to run within the sequence file.
        //
        //	Outputs:
        //		errorOccured:			TRUE or FALSE to represent if an error has occured or not.
        //		errorCode:				The error code if errorOccured is TRUE.
        //		errorMsg:				The error message if errorOccured is TRUE.
        //
        public void ExecuteWithProcessModel(SequenceContext seqContext, string sequenceFileToRun, string processModelName, string sequenceNameParameter,
            out bool errorOccurred, out int errorCode, out String errorMsg)
        {

            string absolutePath;
            bool userCancelled;

            errorOccurred = false;
            errorCode = 0;
            errorMsg = String.Empty;

            try
            {
                // Find the sequence file absolute path based on the calling sequence file's directory.
                seqContext.Engine.FindFile(sequenceFileToRun, out absolutePath, out userCancelled, FindFilePromptOptions.FindFile_PromptHonorUserPreference,
                    FindFileSearchListOptions.FindFile_AddDirToSrchList_Ask, false, seqContext.SequenceFile);

                // Locate and open the sequence file contianing the sequence to be executed.
				SequenceFile targetSeqFile = seqContext.Engine.GetSequenceFileEx(absolutePath);

                // Locate and open the process model to be used for this execution.
				SequenceFile processModel = seqContext.Engine.GetSequenceFileEx(processModelName, GetSeqFileOptions.GetSeqFile_FindFile, TypeConflictHandlerTypes.ConflictHandler_Error);

                // Launch a new execution of the sequence file using the specified process model.
				// The SequenceNameParameter is the name of the process model entry point
				Execution currentExecution = seqContext.Engine.NewExecution(targetSeqFile, sequenceNameParameter, processModel, false, 0);

                // Wait for the execution to end.
                currentExecution.WaitForEndEx(-1, true, Type.Missing, seqContext);

                // Release the process model opened previously.
                seqContext.Engine.ReleaseSequenceFileEx(processModel);
                
                // Release the sequence file opened previously.
                seqContext.Engine.ReleaseSequenceFileEx(targetSeqFile, 0);


            }
            catch (COMException e)
            {
                errorOccurred = true;
                errorMsg = e.Message;
                errorCode = e.ErrorCode;
            }
        }

        //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        //
        //	This method executes a sequence with sequence parameters.
        //
        //	Inputs:
        //		seqContext: 			The sequence context.
        //		sequenceFileToRun:		The name of the sequence file to run located in the TestStand search directories.
        //		sequenceNameParameter:	The name of the sequence to run within the sequence file.
        //		numberValue:			A number to be set as a parameter and displayed to the user.
        //		stringValue:			A string to be set as a parameter and displayed to the user.
        //
        //	Outputs:
        //		errorOccured:			TRUE or FALSE to represent if an error has occured or not.
        //		errorCode:				The error code if errorOccured is TRUE.
        //		errorMsg:				The error message if errorOccured is TRUE.
        //
        public void ExecuteWithSequenceParameters(SequenceContext seqContext, string sequenceFileToRun, string sequenceNameParameter, double numberValue,
            string stringValue, out bool errorOccurred, out int errorCode, out String errorMsg)
        {

            string absolutePath;
            bool userCancelled;

            errorOccurred = false;
            errorCode = 0;
            errorMsg = String.Empty;

            try
            {
                // Find the sequence file absolute path based on the calling sequence file's directory.
                seqContext.Engine.FindFile(sequenceFileToRun, out absolutePath, out userCancelled, FindFilePromptOptions.FindFile_PromptHonorUserPreference,
					FindFileSearchListOptions.FindFile_AddDirToSrchList_Ask, false, seqContext.SequenceFile);

                // Locate and open the sequence file contianing the sequence to be executed.
				SequenceFile targetSeqFile = seqContext.Engine.GetSequenceFileEx(absolutePath);


                // Create a PropertyObject with a subproperty for each parameter.  Subproperties must be in the same order as parameters appear in the sequence.
				PropertyObject sequenceParameters = seqContext.Engine.NewPropertyObject(PropertyValueTypes.PropValType_Container, false, "", 0);
				sequenceParameters.SetValNumber("NumberVal", PropertyOptions.PropOption_InsertIfMissing, numberValue);    // Use PropOption_InsertIfMissing to create the subproperty if it does not exist.
				sequenceParameters.SetValString("StringVal", PropertyOptions.PropOption_InsertIfMissing, stringValue);

                // Launch a new execution of the sequence.
				Execution currentExecution = seqContext.Engine.NewExecution(targetSeqFile, sequenceNameParameter, null, false, 0, sequenceParameters);

                // Wait for the execution to end.
                currentExecution.WaitForEndEx(-1, true, Type.Missing, seqContext);

                // Release the sequence file opened previously.
                seqContext.Engine.ReleaseSequenceFileEx(targetSeqFile, 0);

            }
            catch (COMException e)
            {
                errorOccurred = true;
                errorMsg = e.Message;
                errorCode = e.ErrorCode;
            }
        }
    }
}
